src/tools/pkcs15-tool.c

```diff
@@ -635,6 +637,11 @@ static void print_prkey_info(const struct sc_pkcs15_object *obj)
 	printf("\tAccess Flags   : [0x%02X]", prkey->access_flags);
 	print_key_access_flags(prkey->access_flags);
 	printf("\n");
+	printf("\tAlgo_refs      : ");
+	for (i = 0; i< SC_MAX_SUPPORTED_ALGORITHMS; i++) {
+		printf("%s%u", (i == 0) ? "" : ", ", prkey->algo_refs[i]);
```

Comment:
Can we print just the non-empty ones (if we are use they are initialized)? This create long list of zeroes in my outputs.

Reply:
Main problem is how ASN.1 is parsed.  The count of actually present algo_refs is lost. A zero value may be valid. 
`types.h`  has:
```
 53 /* When changing this value, pay attention to the initialization of the ASN1
 54  * static variables that use this macro, like, for example,
 55  * 'c_asn1_supported_algorithms' in src/libopensc/pkcs15.c,
 56  * src/libopensc/pkcs15-prkey.c and src/libopensc/pkcs15-skey.c
 57  * `grep "src/libopensc/types.h SC_MAX_SUPPORTED_ALGORITHMS  defined as"'
 58  */
 59 #define SC_MAX_SUPPORTED_ALGORITHMS     16
```
if the defineis changed, you have to change the ASN.1 in `pkcs15.c` and `pkcs15-skey.c` to make sure the number of line like this: 
```
{ "algorithmInfo", SC_ASN1_STRUCT, SC_ASN1_TAG_SEQUENCE | SC_ASN1_CONS, SC_ASN1_OPTIONAL, NULL,      NULL },
```
matches the define. Found this when changing limit to 16. One card had 8, the old limit, and did not know if we dropped some. 

SC_PKCS15_MAX_ACCESS_RULES looks like it has the same issue. 

These are all arrays of the same entry. OpenSC needs something like: 
https://www.obj-sys.com/docs/acv60/CCppUsersGuide/CCppUsersGuide27.html

I did not change the ASN1, but would like to as a separate PR. It would have one template  entry that is reused, while counting how many times it is invoked. So one ends up with an array (using realloc) and a count  or a list. 

Could address the pkcs15-tool.c then?

Reply:
Having a dynamic allocation would be indeed better. Thinking about short-term local solution, we could just avoid writing duplicate values, which would make this particular list of 16 zeroes just one zero without the lost of informative value?